### DISCUSSION TITLE
Unified program options in pyloos

### DISCUSSION THREAD

**Discussion by agrossfield** [GrossfieldLab/loos#43] [Ideas]
Hi all,

I've been thinking about making the process of doing command line options in pyloos simpler, in a manner similar to what we did for the c++ tools (put a wrapper on top of the module we're using to process the options, so that you can import groups of commonly used options.

I made a preliminary implementation in a new branch, python-options. I don't want to make a pull request because it's nowhere near ready to be merged. Specifically, there's an options.py module in the pyloos directory (not Packages/PyLOOS), and an example of usage in Packages/PyLOOS/cylindrical-density.py

The biggest ugliness I see is in handling --fullhelp options. Right now, I still need to define the class in the application, because otherwise I can't see how to get the application-specific function embedded, and that's a pretty ugly thing. 

Anyway, I'd appreciate comments and (helpful) suggestions. @tromo ? @lgsmith ?

-- Alan

**Comment by lgsmith** [comment]
One of the reasons that I like argparse is because it feels a lot like the options framework we have in c++ but simplified. I must admit I don't know what the `click` that @tromo mentioned is. I would want it to be as similar to PO as possible, because at least for me learning argparse helped me learn PO.

I would have either done this mostly like you're doing it (but note, as a module I don't think the line `parser = argparse.ArgumentParser()` Is needed, since the user would have to write `loos.options.parser` to access it anyway, and it's just an empty parser), or I would have done the python subclass thing and made a class containing a parser with all the methods you're currently defining as functions.

It seems to me like subclassing parser with something that has a bunch of loos specific methods is a little nicer because then the pyloos user doesn't have to remember to import argparse and instantiate an empty parser on their own. In that way perhaps it's a bit more similar to our subclassing of PO. As written above, that homology between the C++ and Python codes is pretty virtuous, and while the python should strive to be cleaner and more pythonic I think it should not be so at the expense of totally losing the homology. As a corollary, if two things are roughly equally simple in python but one is more homologous to the loos C++ library we should prefer that.

As to how to do fullhelp, well, all fullhelp is in python tools is a triple quote string, so you'd just make a parser method that stores a string for printing upon calling fullhelp. The class's init method can start with that string as an empty string, or perhaps (to warn people that they need to feed it a string) something like `'there's no fullhelp yet'`. You want the triple-quote string to be in the tool, so that autodocs tools that look for those can grab it.

**Reply by agrossfield** [reply]
just on the fullhelp: that doesn't work. --fullhelp is currently done as a custom Action.

Because argparse takes a class (not an object of that type), once the fullhelp option is set up there's no way to change the message. At the same time, there's no way to pass an extra argument in, so you can't define a fullhelp string and then pass it along. Or at least I couldn't find one -- I spent a while messing with it, and couldn't come up with a way to do it. I was even desperate enough to try using a global variable, but even that doesn't work that way (a module that's imported can't access globals from the module that did the importing).

**Reply by lgsmith** [reply]
Wait you are saying you can't have your custom subclass have a fullhelpstring as part of its state, and give your parser an option that is just a boolswitch that prints an argument string (self.fullhelp) then calls exit(0)?

**Reply by lgsmith** [reply]
Something like this:

```
import argparse as ap

class po:
    def __init__(self, parser: ap.ArgumentParser):
        self.parser = parser
        self.fullhelp = 'fullhelp'
    def add_fullhelp(self, fullhelp):
        self.fullhelp = fullhelp
        self.parser.add_argument('--fullhelp', action='store_true', default=False)
    def parse_args(self):
        args = self.parser.parse_args()
        # postprocess args here; error checks, checks for needed behavior
        if args.fullhelp:
            print(self.fullhelp)
            exit(0)
        return args
```

**Reply by agrossfield** [reply]
That's different from how it currently works. I think what you're describing would work, although it's not the way argparse is intended to be used (the way we're doing it is, where you create a custom action).

**Reply by agrossfield** [reply]
actually, there's no reason for that to be a subclass. In your example, `po` just contains an `ap.ArgumentParser` object.

**Reply by lgsmith** [reply]
Yes, I decided that was simpler and wrote it as a wrapper instead.

**Comment by lgsmith** [comment]
Yes. I decided that was simpler.

**Comment by agrossfield** [comment]
I pushed a second version today, as a RFC. Thoughts?

**Reply by lgsmith** [reply]
Noticed a small booboo in the argv length check (I believe). I just fixed it and pushed it back. This is basically what I was thinking about, although the length check that prints the parser autohelp statement seems like it belongs in the parse_args statement rather than init.

Something that may also be nice, and that I kind of wish the PO had (perhaps I should put it in) is sort of a verbose mode where it pukes up all the options it thinks it got to std error in the case of any kind of runtime issue or segfault. The pythonic equivalent of that (and of writing the 'header string, that almost all loos tools write) probably belongs in the argparse wrapper even though in the CPP they're separate.

**Comment by agrossfield** [comment]
You’re right, and I’ll add a generateHeader option.

________________________________
Dr. Alan Grossfield
Dept of Biochemistry and Biophysics
University of Rochester Medical Center
Phone: 585 276 4193
http://membrane.urmc.rochester.edu
https://orcid.org/0000-0002-5877-2789
Pronouns: He/his



From: lgsmith ***@***.***>
Reply-To: GrossfieldLab/loos ***@***.***>
Date: Thursday, April 8, 2021 at 2:55 PM
To: GrossfieldLab/loos ***@***.***>
Cc: "Grossfield, Alan" <Alan_Grossfield@URMC.Rochester.edu>, Author ***@***.***>
Subject: [EXT] Re: [GrossfieldLab/loos] Unified program options in pyloos (#43)


Noticed a small booboo in the argv length check (I believe). I just fixed it and pushed it back. This is basically what I was thinking about, although the length check that prints the parser autohelp statement seems like it belongs in the parse_args statement rather than init.

Something that may also be nice, and that I kind of wish the PO had (perhaps I should put it in) is sort of a verbose mode where it pukes up all the options it thinks it got to std error in the case of any kind of runtime issue or segfault. The pythonic equivalent of that (and of writing the 'header string, that almost all loos tools write) probably belongs in the argparse wrapper even though in the CPP they're separate.

**Reply by lgsmith** [reply]
I'm waffling about the subclass vs wrapper thing, again. It would be kind of nice not to have to write `loosparser.parser.add_argument(argjunk)`. It would be some added familiarity for people to not have to do that, I think.

I will try to write a version of the exact same options code as a subclass this evening, if I get a free minute. Then you can look at them both and decide.

**Comment by agrossfield** [comment]
RFC: take a look at the latest iteration of options.py
To get an idea of what it feels like in use, I wrote a new tool called contact_distance.py (basically, rmsds but comparing the matrix of residue-residue contacts instead of the RMSD). I'm moderately happy with it, but I want some feedback before I merge.

**Comment by agrossfield** [comment]
I'm leaning toward merging this in but not going back and retrofitting the other tools right away. Any objections?

**Reply by agrossfield** [reply]
I've made a pull request ( #49 ). I'll merge it next week if there are no objections
